Configuration
=============

The following provides an overview of all available flags in the configuration file to use the high level API and run jobs.

Always required
---------------

.. table::
   :class: fixed-table main-config-table

   +----------------------------+--------------------------------------------------------------------------------------------------------------------+
   | Parameter                  |                             Description                                                                            |
   +============================+====================================================================================================================+
   | type                       | The type of job; can be one of "CollisionEnergCalibration", "SpectralLibraryGeneration" or "Rescoring"             |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------+
   | tag                        | Optional mass tag Can be "tmt", "tmtpro", "itraq4" or "itraq8"; default is ""                                      |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------+
   | models                     | Contains information about the used models for peptide property prediction (see following 2 nested parameters)     |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------+
   |     intensity              | Name of the model used for fragment intensity prediction                                                           |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------+
   |     irt                    | Name of the model used for indexed retention time prediction                                                       |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------+
   | prediction_server          | Server and port for obtaining peptide property predictions; default: "koina.wilhelmlab.org:443"                    |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------+
   | ssl                        | Use ssl when making requests to the prediction server, can be true or false; default = true                        |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------+
   | output                     | Path to the output folder (relative to the location of the config file); default = "./"                            |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------+

Required for CE calibration and rescoring
-----------------------------------------

.. table::
   :class: fixed-table lib-rescore-config-table

   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | Parameter                  |                             Description                                                                                                                                                                                                                                                    |
   +============================+============================================================================================================================================================================================================================================================================================+
   | inputs                     | Contains information about inputs and the type of the inputs (see following 4 nested parameters)                                                                                                                                                                                           |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     search_results         | Path to directory or file containing the search results                                                                                                                                                                                                                                    |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     search_results_type    | Format description for search results; can be "Maxquant", "Msfragger", "Mascot", "Sage" or "`Internal <./internal_format.html>`_"; default = "Maxquant"                                                                                                                                    |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     spectra                | Path to directory or file containing mass spectrometry results                                                                                                                                                                                                                             |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     spectra_type           | Format description for files containing spectra; can be "raw", "mzml", "d" or "hdf"; default = "raw"; in case of mixed mzML/RAW input, select "raw"; in case of mixed d/hdf input select "d"                                                                                               |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | numThreads                 | Number of raw/mzml files processed in parallel (parallelisation on file level); more processes than files has no effect and should be avoided; for spectral library generation, the number of parallel prediction processes, needs to be balanced with batchsize in this case; default = 1 |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | thermoExe                  | Path to ThermoRawFileParser executable; needed if spectra are provided in RAW format; default "ThermoRawFileParser.exe"                                                                                                                                                                    |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | massTolerance              | Defines the allowed tolerance between theoretical and experimentally observered fragment mass during peak annotation; default = 20 (FTMS), 40 (TOF), 0.35 (ITMS)                                                                                                                           |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | unitMassTolerance          | Defines the measure of tolerance, either "da" or "ppm"; default = da (mass analyzer is ITMS), ppm (mass analyzer is FTMS or TOF)                                                                                                                                                           |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | ce_alignment_options       | Contains settings for collision energy alignment                                                                                                                                                                                                                                           |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     ce_range               | Min and max collision energy (end-exclusive) used for calibration; i.e. (5,10) tests every CE from 5 to 9. Default is (19,50)                                                                                                                                                              |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     use_ransac_model       | Boolean that determines whether or not to use a ransac regression model for calibration refinement. This is recommended for timsTOF data. Default is false.                                                                                                                                |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

Additional parameters required for spectral library generation
--------------------------------------------------------------

.. table::
   :class: fixed-table lib-config-table

   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | Parameter                  |                             Description                                                                                                                                          |
   +============================+==================================================================================================================================================================================+
   | inputs                     | Contains information about additional inputs required for spectral library generation and the type of the inputs (see following 2 nested parameters)                             |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     library_input          | Path to fasta file for in-silico digestion (also see the required parameters in the following section) or an existing output file from a digestion                               |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     library_input_type     | Library input type description; can be "fasta" if an additional in-silico digestion is required or "`peptides <./peptides_format.html>`_"                                        |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | spectralLibraryOptions     | Contains information about additional settings required for spectral library generation and what to save to disk (see following 7 nested parameters)                             |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     fragmentation          | Method used for fragmentation; can be "HCD" or "CID"; default = ""                                                                                                               |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     collisionEnergy        | The collision energy for which the library should be created; default = 30                                                                                                       |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     precursorCharge        | The precursor charges for which the library should be created, can be a list or single number; default = [2,3]                                                                   |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     minIntensity           | The minimal relative intensity threshold for peaks, everything below is not saved, this can help reducing the library size; default = 5e-4                                       |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     batchSize              | Number of peptides for which predictions are retrieved at once before writing, larger batches result in higher memory peaks, needs to be balanced with n_threads, default = 10000|
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     format                 | Output format of the generated spectral library; can be "spectronaut" or "msp"; default = "msp"                                                                                  |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

Additional parameters required for in-silico digestion
------------------------------------------------------

.. table::
   :class: fixed-table digest-config-table

   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | Parameter                  |                             Description                                                                                                                            |
   +============================+====================================================================================================================================================================+
   | fastaDigestOptions         | Contains specific settings for the in-silico digestion of a provided fasta file (see following 8 nested parameters)                                                |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     digestion              | Digestion mode; can be "full", "semi" or None; default = "full"                                                                                                    |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     missedCleavages        | Number of allowed missed cleavages used in the search engine for generation of the provided search results; default = 2                                            |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     minLength              | Minimum peptide length allowed used in the search engine for generation of the provided search results; default = 7                                                |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     maxLength              | Minimum peptide length allowed used in the search engine for generation of the provided search results; default = 60                                               |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     enzyme                 | Name of the enzyme used in the search engine; default = "trypsin"                                                                                                  |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     specialAas             | Special amino acids for decoy generation; default = "KR"                                                                                                           |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   |     db                     | Defines whether the digestion should contain only targets, only decoys or both (concatenated); can be "target", "decoy" or "concat"; default = "concat"            |
   +----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------+

Additional parameters required for rescoring
--------------------------------------------

.. table::
   :class: fixed-table

   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | Parameter                  |                             Description                                                                                                                                          |
   +============================+==================================================================================================================================================================================+
   | fdr_estimation_method      | Method used for target / decoy separation and FDR estimation on PSM and peptide level: "percolator" or "mokapot"; default = "mokapot"                                            |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | allFeatures                | Defines whether all features should be used; can be True or False; default = False                                                                                               |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
   | regressionMethod           | Regression method for curve fitting (mapping from predicted iRT values to experimental retention times); can be "spline", "lowess", or "logistic"; default = "spline"            |
   +----------------------------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
